### ISSUE TITLE
Serve Sprites

### ISSUE THREAD

**Issue by indus** [klokantech/tileserver-gl#131]
I´m unable to serve sprites. Documentation & code suggests to set up a sprites path in the config file but that looks like a deadend.
Thats https://github.com/klokantech/tileserver-gl/blob/master/src/server.js#L87 the last use of sprites that I can find.

**Comment by strech345** [comment]
same problem. I have copy 4 files (sprite.json, sprite.png, sprite@2x.json, sprite@2x.png ) to sprites/ , sprites/<styleId> , styles/<StyleId>/,  styles/<StyleId>/styles. I didn't get it to work. Which folder should be used for sprites and how should be the conf and style?

**Comment by strech345** [comment]
I found now the problem.
like for mbtiles and fonds i set the domain of the server in the style that doesn't work.
if i only set the local path "\<style-id\>/sprite" it works.
4 Files are in "sprites/\<style-id>\/"

**Comment by petrsloup** [comment]
Make sure you are using the latest version (1.5.0).

If you run the tileserver-gl with `--verbose` you can see the generated config file and base your configuration on that + prebuilt styles from https://www.npmjs.com/package/tileserver-gl-styles

**Comment by karakostis** [comment]
I am using the latest version of TileServer GL. I am trying to replace the circles with symbols. I my config.json file I have the following line:

`"sprites": "sprites",`

Also in the root folder of the tileserver, I have a folder (sprites) with my sprites (json/png etc.).

How do I display this sprite on my map? In my style.json I try the following:

```
{
      "id": "wld_airports",
      "type": "symbol",
      "source": "wld_airports",
      "source-layer": "wld_airports",
      "paint": {},
      "layout": {
        "icon-image": "airport-15"
      },
      "minzoom": 1,
      "maxzoom": 20
    },
```

But nothing is displayed. When I try to display them as circles (as below), I see my data:

```
{
      "id": "wld_airports",
      "type": "circle",
      "source": "wld_airports",
      "source-layer": "wld_airports",
      "paint": {
        "circle-color": "rgba(199, 9, 9, 1)",
        "circle-radius": 9,
        "circle-stroke-width": 2,
        "circle-stroke-color": "rgba(37, 63, 224, 1)"
      },
      "minzoom": 1,
      "maxzoom": 20
    },
```

**Comment by alexdawn** [comment]
I can serve sprites for only the style specific local path: `<style-id>/sprite` regardless of config setting or using {styleJsonFolder} in the style.json the path remains `<style-id>/sprite` and I get a 404 error for the sprite resource unless they are in that local path. How do I set it up so I can serve a common set of sprites to many styles?

**Comment by danger89** [comment]
I still really think there is a problem with tileserver-gl. Since I can only serve my sprites when the sprite.png/json files etc. are located within the styles directory. While I would like to store the sprites in.. the .. sprites directory of course.

Putting it in `styles` directory I get a 200 OK, and indeed the sprites are displayed (eg. when using OSM bright style). Putting the sprite files in the `sprites` directory, I keep getting 404. I also tried to put the files within `sprites/<style_id>` (like `sprites/osm-bright` in my case). I keep getting 404's.. this is really crazy..

Please consider to reopen.

**Comment by zeevogels** [comment]
I agree, please consider reworking the tileserver so that it's also possible to only serve sprites (under /sprites), without a style. This way a sprite can be reused across styles.

**Comment by cybersyntactics** [comment]
For anyone who happens to come across this issue in the future, it appears the way it is working is that tileserver-gl is re-writing several things in the style json file and serving things at different locations than their directory structure would suggest. 

Particularly:
- <style_path>/<style_id>.json will actually be served at <domain>/style/<style_id>/style.json
- <sprite_path>/<sprite_id>.<json | png> (and the @2x equivelents) will be served at <domain>/styles/<style_id>/sprite.json, etc. 
- The `sprites` field inside the style json, if using a local path `<sprite_id>` will be re-written to `<domain>/styles/<style_id>/sprite`

**Comment by jesperp** [comment]
This had me confused for a couple of hours, so here is my understanding of this.

From your config.json, the value of `options.paths.sprites` is joined with `sprite` from your style.json. So if I had a config.json like this..

```json
{
  "options": {
    "paths": {
      "root": "",
      "fonts": "fonts",
      "sprites": "sprites",
      "styles": "styles",
      "mbtiles": "data"
    },
  "...": "..."
}
```

And a style:
```json
{
  "name": "My Style",
  "version": 8,
  "center": [14.5097, 35.8893],
  "zoom": 14,
  "glyphs": "{fontstack}/{range}.pbf",
  "sprite": "Malta/sprite",
  "...": "..."
}
```

The joined path will be `sprites/Malta/sprite`. Finally (this was the confusing part for me), the files that need to exist on your disk from this joined path are then:

```
sprites/Malta/sprite.json
sprites/Malta/sprite.png
sprites/Malta/sprite@2x.json
sprites/Malta/sprite@2x.png
```

This means that if I edit `sprite` in my style.json to say `"stuff"` it would look for files `sprites/stuff.json`, etc..

**Comment by druwan** [comment]
Thank you @jesperp, I've been banging my head for almost 2 days!

**Comment by julienmey21** [comment]
It took me days to figure out, I have my sprite files in a sprites folder in the root directory of Tileserver (using Docker):

My config.json:

 "options": {
      "paths": {
        "root": "/usr/src/app/node_modules/tileserver-gl-styles",
        "fonts": "fonts",
        "sprites": "/data/sprites",
        "styles": "/data",
        "mbtiles": "/data"
      }

My style.json: 

 "sprite": "/sprite",
  "glyphs": "/fonts/{fontstack}/{range}.pbf",

Basically, the path in style.json will be added at the end of the one in config.json in other words: /data/sprites/sprite, which works for me

**Comment by k-i-o** [comment]
> It took me days to figure out, I have my sprite files in a sprites folder in the root directory of Tileserver (using Docker):
> 
> My config.json:
> 
> "options": { "paths": { "root": "/usr/src/app/node_modules/tileserver-gl-styles", "fonts": "fonts", "sprites": "/data/sprites", "styles": "/data", "mbtiles": "/data" }
> 
> My style.json:
> 
> "sprite": "/sprite", "glyphs": "/fonts/{fontstack}/{range}.pbf",
> 
> Basically, the path in style.json will be added at the end of the one in config.json in other words: /data/sprites/sprite, which works for me

and with that config ur call is https://domain.xxx/sprite.json and it works?

i put in /data/sprites 2 files, sprite.json and sprite.png with the collage of all symbols, two CORRECT files bc i took them from a working project of another person.

my config:
```{
    "options": {
        "paths": {
            "root": "/data",
            "fonts": "fonts",
            "sprites": "sprites",
            "styles": "styles",
            "mbtiles": "/data"
        },
        "serveStaticMaps": true,
        "formatQuality": {
            "png": 90,
            "jpeg": 90,
            "webp": 90
        },
        "maxSize": 8192,
        "pbfAlias": "pbf"
    },
    "styles": {
        "basic": {
            "style": "basic-style.json",
            "tilejson": {
                "bounds": [60.8786, 23.32744, 77.83562, 37.09664]
            }
        }
    },
    "data": {
        "v3": {
            "mbtiles": "/data/planet.mbtiles"
        }
    }
}
```

my basic style:
```{
  "version": 8,
  "name": "Basic preview",
  "metadata": {
    "openmaptiles:version": "3.x",
    "maputnik:renderer": "mlgljs"
  },
  "sources": {
    "openmaptiles": {
      "type": "vector",
      "url": "https://domain.xxx/data/v3.json"
    }
  },
  "sprite": "https://domain.xxx/styles/basic/sprite",
  "glyphs": "https://domain.xxx/fonts/{fontstack}/{range}.pbf",
  "layers": [
    {
      "id": "background",
      "type": "background",
      "filter": ["all"],
      "layout": {
        "visibility": "visible"
      },
      "paint": {
        "background-color": "hsl(47, 26%, 88%)"
      }
    },
    {
      "id": "water",
      "type": "fill",
      "source": "openmaptiles",
      "source-layer": "water",
      "filter": [
        "all",
        ["==", "$type", "Polygon"],
        ["!=", "intermittent", 1],
        ["!=", "class", "lake"],
        ["!=", "class", "river"],
        ["!=", "class", "pond"],
        ["!=", "class", "swimming_pool"]
      ],
      "layout": {
        "visibility": "visible"
      },
      "paint": {
        "fill-color": "hsl(205, 56%, 73%)",
        "fill-antialias": false,
        "fill-pattern": "wave"
      }
    }
  ],
  "id": "basic"
}```

NOTE:
- fonts works
- styles works
- preview works
- http://domain.xxx/styles/basic/#4.07/71.98/-39.7 <- there i see the map preview
- http://domain.xxx/styles/basic/style.json <- there i see the map style
- http://domain.xxx/styles/basic/sprite.json <- there i see 'Bad Sprite ID or Scale'
- http://domain.xxx/sprite.json <- there i see 'Cannot GET /sprite.json'
- http://domain.xxx/sprites/sprite.json <- there i see 'Cannot GET /sprites/sprite.json'

i don't rly know how to find the sprites

**Comment by k-i-o** [comment]
Now, someone know how to fill-pattern the ground as well as the water? Because the ground is the background and does not have a fill pattern.
